File: week2/week2.py
# Unit test
# Logging interface
# HW

def task1():
    def find_and_print(messages):
        over_17_name = []
        for name,sentence in messages.items():
            # Skip the negative sentences because it is hard to judge
            if sentence.find("not")!=-1 and sentence.find("n't")!=-1 : continue
            # Check the ages
            if sentence.find("years old")!=-1:
                tokens = sentence.split()
                for token in tokens:
                    if token.isdigit() and int(token) > 17:
                        over_17_name.append(name)
            # Check the role
            elif sentence.find("college student")!=-1 or sentence.find("university student")!=-1:
                over_17_name.append(name)
            # Check the legal age
            elif sentence.find("legal age")!=-1:
                over_17_name.append(name)
            # Check the qualification of voting
            elif sentence.find("vote")!=-1:
                over_17_name.append(name)

        print("[{step}] - possible older than 17 years old name : {over_17_name}".format(step="task1", over_17_name=over_17_name))
    find_and_print({
        "Bob":"My name is Bob. I'm 18 years old.",
        "Mary":"Hello, glad to meet you.",
        "Copper":"I'm a college student. Nice to meet you.",
        "Leslie":"I am of legal age in Taiwan.",
        "Vivian":"I will vote for Donald Trump next week",
        "Jenny":"Good morning."
    })

def task2():
    def calculate_sum_of_bonus(data):
        name2bonus = {}
        max_bonus = 10000
        for employees_info in data.values():
            for employee in employees_info:
                # Bonus :
                #   max_bonus * (performance_gain + salary_gain)
                #   salary_gain : salary/max_bonus
                #   performance_gain : based on different performance

                # Parse salary
                salary_str = "".join(str(employee["salary"]).split(","))
                salary_gain = 0.01*(float(salary_str) if str(salary_str).find("USD")==-1 else float(salary_str.split("USD")[0])*30) \
                            /max_bonus
                # Parse performance
                performance_gain = 0.4 # Default is 20%
                if employee["performance"].find("above"):
                    performance_gain = 0.7
                elif employee["performance"].find("below"):
                    performance_gain = 0.2
                # Add name and bonus
                final_gain = 1.0 if (performance_gain + salary_gain)>1.0 else (performance_gain + salary_gain)
                name2bonus[employee["name"]] = max_bonus * final_gain
                # Sanity check
                assert(name2bonus[employee["name"]] <= max_bonus)
        # Print
        for name,bonus in name2bonus.items():
            print("[{step}] - name : {name}, bonus : {bonus}".format(step="task2", name=name, bonus=bonus))

    calculate_sum_of_bonus(
    {
        "employees":
        [
            {
                "name":"John",
                "salary":"1000USD",
                "performance":"above average",
                "role":"Engineer"
            },
            {
                "name":"Bob",
                "salary":60000,
                "performance":"average",
                "role":"CEO"
            },
            {
                "name":"Jenny",
                "salary":"50,000",
                "performance":"below average",
                "role":"Sales"
            }
        ]
    }
    ) # call calculate_sum_of_bonus function

def task3():
    def func(*data):
        middle_name_list = [name[1] for name in data]
        middle_name_cnt = {}
        # Construct the dictionary based on the key : middle name, value : count of middle name
        for middle_name in middle_name_list:
            if middle_name in middle_name_cnt:
                middle_name_cnt[middle_name] += 1
            else:
                middle_name_cnt[middle_name] = 1
        ans_name = "沒有"
        # Iterate the name list to get the corresponding full name (There is only one full name corresponding to unique middle name)
        for middle_name,value in middle_name_cnt.items():
            if value==1:
                for name in data:
                    if name[1]==middle_name:
                        ans_name = name
        print("[{step}] - name = {ans_name}".format(step="task3", ans_name=ans_name))
    func("彭⼤牆", "王明雅", "吳明") # print 彭⼤牆
    func("郭靜雅", "王立強", "林靜宜", "郭立恆", "林花花") # print 林花花
    func("郭宣雅", "林靜宜", "郭宣恆", "林靜花") # print 沒有

def task4():
    def get_number(index):
        # 0 4 3 7 6 10 9 13 12 16 15
        print("[{step}] - index : {index}, ans : {ans}".format(step="task4", index=index, ans=4*(int(index/2)+index%2)-int(index/2)))
    get_number(1) # print 4
    get_number(5) # print 10
    get_number(10) # print 15

def task5():
    def find_index_of_car(seats, status, number):
        # Construct a list with available seat => (index, value)
        idx_with_seats = [(index, seats[index]) for index, value in enumerate(status) if value==1]
        # Check 
        #   1. the available seats is larger thant the required number
        #   2. the difference is the smallest(most fitted)
        difference = 2**64-1
        most_fit_train_index = -1
        for index,seats_num in idx_with_seats:
            if (seats_num - number) >= 0 and (seats_num - number) < difference:
                difference = (seats_num - number)
                most_fit_train_index = index
        print("[{step}] - most fitted train index : {most_fit_train_index}".format(step="task5", most_fit_train_index=most_fit_train_index))
        
    find_index_of_car([3, 1, 5, 4, 2], [0, 1, 0, 1, 1], 2) # print 4
    find_index_of_car([1, 0, 5, 1, 3], [0, 1, 0, 1, 1], 4) # print -1
    find_index_of_car([4, 6, 5, 8], [0, 1, 1, 1], 4) # print 2

if __name__=="__main__":
    print("="*100)
    task1()
    print("="*100)
    task2()
    print("="*100)
    task3()
    print("="*100)
    task4()
    print("="*100)
    task5()

File: week2/week2.js
function task1(){
    function findAndPrint(messages){
        over_17_name = []
        for(const [name,sentence] of Object.entries(messages)){
            // Skip the negative sentences because it is hard to judge
            if(sentence.search("not")!=-1 && sentence.search("n't")!=-1) continue;
            // Check the ages
            if(sentence.search("years old")!=-1){
                tokens = sentence.split(' ')
                for(token of tokens){
                    let number = parseInt(token);
                    // console.log(number);
                    if(number > 17){
                        over_17_name.push(name);
                        break;
                    }
                }
            }
            // Check the role
            else if(sentence.search("college student")!=-1 || sentence.search("university student")!=-1){
                over_17_name.push(name)
            }
            // Check the legal age
            else if(sentence.search("legal age")!=-1){
                over_17_name.push(name)
            }
            // Check the qualification of voting
            else if(sentence.search("vote")!=-1){
                over_17_name.push(name)
            }
        }
        console.log(`[task1] - possible older than 17 years old name : ${over_17_name}`)
    }
    findAndPrint({
        "Bob":"My name is Bob. I'm 18 years old.",
        "Mary":"Hello, glad to meet you.",
        "Copper":"I'm a college student. Nice to meet you.",
        "Leslie":"I am of legal age in Taiwan.",
        "Vivian":"I will vote for Donald Trump next week",
        "Jenny":"Good morning."
    });
}

function task2(){
    function calculateSumOfBonus(data){
        name2bonus = {};
        max_bonus = 10000;
        for(employees_info of Object.values(data)){
            for(employee of employees_info){
                // Bonus :
                //   max_bonus * (performance_gain + salary_gain)
                //   salary_gain : salary/max_bonus
                //   performance_gain : based on different performance
                // Parse salary
                let salary_str = (String(employee["salary"]).split(",")).join('');
                let salary_gain = 0.01*(String(salary_str).search("USD")==-1 ? parseFloat(salary_str) : (parseFloat(String(salary_str).split('USD')[0])*30))/max_bonus;
                // Parse performance
                let performance_gain = 0.4 // Default is 40%
                if(employee["performance"].search("above"))
                    performance_gain = 0.7;
                else if(employee["performance"].search("below"))
                    performance_gain = 0.2;
                // Add name and bonus
                let final_gain = (performance_gain + salary_gain)>1.0 ? 1.0 : (performance_gain + salary_gain);
                name2bonus[employee["name"]] = max_bonus * final_gain;
            }
        }
        for(const [name, bonus] of Object.entries(name2bonus))
            console.log(`[task2] - name = ${name}, bonus : ${bonus}`);
    }
    
    calculateSumOfBonus(
        {
            "employees":[
                {
                    "name":"John",
                    "salary":"1000USD",
                    "performance":"above average",
                    "role":"Engineer"
                },
                {
                    "name":"Bob",
                    "salary":60000,
                    "performance":"average",
                    "role":"CEO"
                },
                {
                    "name":"Jenny",
                    "salary":"50,000",
                    "performance":"below average",
                    "role":"Sales"
                }
            ]
        }
    ); // call calculateSumOfBonus function
}

function task3(){
    function func(...data){
        middle_name_list = []
        middle_name_cnt = {}
        for(const name of data)
            middle_name_list.push(name[1]);
        // Construct the dictionary based on the key : middle name, value : count of middle name
        for(const middle_name of middle_name_list){
            if(middle_name_cnt.hasOwnProperty(middle_name))
                middle_name_cnt[middle_name] += 1;
            else
                middle_name_cnt[middle_name] = 1;
        }
        ans_name = "沒有"
        // Iterate the name list to get the corresponding full name (There is only one full name corresponding to unique middle name)
        for(const[middle_name,value] of Object.entries(middle_name_cnt)){
            if(value==1){
                for(const name of data)
                    if(name[1]==middle_name)
                        ans_name = name
            }
        }
        console.log(`[task3] - name = ${ans_name}`);
    }

    func("彭⼤牆", "王明雅", "吳明"); // print 彭⼤牆
    func("郭靜雅", "王立強", "林靜宜", "郭立恆", "林花花"); // print 林花花
    func("郭宣雅", "林靜宜", "郭宣恆", "林靜花"); // print 沒有
}

function task4(){
    function getNumber(index){
        let ans = 4*(Math.floor(index/2)+index%2)-Math.floor(index/2);
        console.log(`[task4] - index : ${index}, ans : ${ans}`);
    }
    getNumber(1); // print 4
    getNumber(5); // print 10
    getNumber(10); // print 15
}

function task5(){
    function findIndexOfCar(seats, status, number){
        idx_with_seats = {};
        for(let i=0 ; i<status.length ; i++){
            if(status[i] == 1){
                idx_with_seats[i] = seats[i];
            }
        }
        difference = 2**64-1;
        most_fit_train_index = -1;
        for(const [index,seats_num] of Object.entries(idx_with_seats)){
            if((seats_num - number) >= 0 &&  (seats_num - number) < difference){
                difference = (seats_num - number);
                most_fit_train_index = index;
            }
        }
        console.log(`[task5] - most fitted train index : ${most_fit_train_index}`);
    }
    findIndexOfCar([3, 1, 5, 4, 2], [0, 1, 0, 1, 1], 2); // print 4
    findIndexOfCar([1, 0, 5, 1, 3], [0, 1, 0, 1, 1], 4); // print -1
    findIndexOfCar([4, 6, 5, 8], [0, 1, 1, 1], 4); // print 2
}

task1();
task2();
task3();
task4();
task5();